pkg/nfd-master/nfd-master.go

```diff
@@ -0,0 +1,337 @@
+	stdoutLogger.Printf("NodeName: '%s'", nodeName)
+
+	// Initialize Kubernetes API helpers
+	helper := apihelper.APIHelpers(apihelper.K8sHelpers{})
+
+	if !m.args.NoPublish {
+		err := updateMasterNode(helper)
+		if err != nil {
+			return fmt.Errorf("failed to update master node: %v", err)
+		}
+	}
+
+	// Create server listening for TCP connections
+	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", m.args.Port))
+	if err != nil {
+		return fmt.Errorf("failed to listen: %v", err)
+	}
+	// Notify that we're ready to accept connections
+	m.ready <- true
+	close(m.ready)
+	time.Sleep(time.Second)
+
+	serverOpts := []grpc.ServerOption{}
+	// Enable mutual TLS authentication if --cert-file, --key-file or --ca-file
+	// is defined
+	if m.args.CertFile != "" || m.args.KeyFile != "" || m.args.CaFile != "" {
+		// Load cert for authenticating this server
+		cert, err := tls.LoadX509KeyPair(m.args.CertFile, m.args.KeyFile)
+		if err != nil {
+			return fmt.Errorf("failed to load server certificate: %v", err)
+		}
+		// Load CA cert for client cert verification
+		caCert, err := ioutil.ReadFile(m.args.CaFile)
+		if err != nil {
+			return fmt.Errorf("failed to read root certificate file: %v", err)
+		}
+		caPool := x509.NewCertPool()
+		if ok := caPool.AppendCertsFromPEM(caCert); !ok {
+			return fmt.Errorf("failed to add certificate from '%s'", m.args.CaFile)
+		}
+		// Create TLS config
+		tlsConfig := &tls.Config{
+			Certificates: []tls.Certificate{cert},
+			ClientCAs:    caPool,
+			ClientAuth:   tls.RequireAndVerifyClientCert,
+		}
+		serverOpts = append(serverOpts, grpc.Creds(credentials.NewTLS(tlsConfig)))
+	}
+	m.server = grpc.NewServer(serverOpts...)
+	pb.RegisterLabelerServer(m.server, &labelerServer{args: m.args, apiHelper: helper})
+	stdoutLogger.Printf("gRPC server serving on port: %d", m.args.Port)
+	return m.server.Serve(lis)
+}
+
+// Stop NfdMaster
+func (m *nfdMaster) Stop() {
+	m.server.Stop()
+}
+
+// Wait until NfdMaster is able able to accept connections.
+func (m *nfdMaster) WaitForReady(timeout time.Duration) bool {
+	select {
+	case ready, ok := <-m.ready:
+		// Ready if the flag is true or the channel has been closed
+		if ready == true || ok == false {
+			return true
+		}
+	case <-time.After(timeout):
+		return false
+	}
+	// We should never end-up here
+	return false
+}
+
+// Advertise NFD master information
+func updateMasterNode(helper apihelper.APIHelpers) error {
+	cli, err := helper.GetClient()
+	if err != nil {
+		return err
+	}
+	node, err := helper.GetNode(cli, nodeName)
+	if err != nil {
+		return err
+	}
+
+	// Advertise NFD version as an annotation
+	addAnnotations(node, Annotations{"master.version": version.Get()})
+	err = helper.UpdateNode(cli, node)
+	if err != nil {
+		stderrLogger.Printf("can't update node: %s", err.Error())
+		return err
+	}
+
+	return nil
+}
+
+// Filter labels if whitelist has been defined
+func filterFeatureLabels(labels *Labels, labelWhiteList *regexp.Regexp) {
```

Comment:
As said in another PR, this function is not used. I'm not sure we may fix it in this PR, but we should open an issue or something for this I think.

Reply:
Argh, yes. Let's handle this separately. I opened #237 where we can discuss it